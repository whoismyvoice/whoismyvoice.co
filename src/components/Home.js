import React from 'react';
import SenateStore from '../stores/SenateStore';
import Settings from '../data/settings.json';

import DataUtils from '../utils/DataUtils';
import ContainerActions from '../actions/ContainerActions';
import cx from 'classnames';
import Results from './Results';

// Components
import BaseComponent from './BaseComponent';
import Circle from './Circle';
import SearchGroup from './SearchGroup';
import CongressmanGroup from './Senator/CongressmanGroup';
import WhiteBorder from './WhiteBorder';

// Styles
import style from './../styles/Home.scss';

class Home extends BaseComponent {
  constructor() {
    super();
    this._bind('_handleChange', '_initializeFullpage', '_destroyFullpage');
    this.state = SenateStore.getMember();
  }

  componentDidMount() {
    if (this.state.did_search) {
      this._initializeFullpage();
    }
    if (process.env.NODE_ENV === 'production' && this.state.settings === null) {
      DataUtils.saveFetchedData();
    }
    SenateStore.addChangeListener(this._handleChange);
  }

  componentWillUnmount() {
    SenateStore.removeChangeListener(this._handleChange);
    this._destroyFullpage();
  }

  _handleChange() {
    this.setState(SenateStore.getMember());
  }

  _initializeFullpage () {
    $('#fullpage').fullpage({
      navigation: false,
      showActiveTooltip: false,
      slidesNavigation: false,
      css3: true,
      autoScrolling: true,
      fitToSection: true,
      easingcss3: 'ease-in',
      loopHorizontal: false,
      keyboardScrolling: false,
      animateAnchor: true,
      recordHistory: true,
      controlArrows: false,
      verticalCentered: false,
      resize: true,
      onLeave: (index, nextIndex) => {
        ContainerActions.identifySection(nextIndex);
      }
    });
  }
  _destroyFullpage() {
    if ($.fn.fullpage.destroy !== undefined) {
      $.fn.fullpage.destroy('all');
    }
  }
  render() {
    const NUMBER_REPRESENTATIVES = this.state.number_representatives,
          SETTINGS = this.state.settings,
          REPRESENTATIVES = this.state.representatives,
          DID_SEARCH = this.state.did_search,
          ERROR = this.state.error,
          ZIP_CODE = this.state.zip_code,
          STATE_FULL = this.state.state_full,
          CURRENT_MEMBER = this.state.current_senator,
          {single_voted_for, single_voted_against} = SETTINGS ? SETTINGS.house : Settings.house,
          {cosponsor_post_text, impact_text, represent} = SETTINGS ? SETTINGS.senate : Settings.senate,
          {chamber, bill_desc} = SETTINGS ? SETTINGS : Settings,
          MEMBER = chamber === 'senate' ? 'senator' : 'representative';

    let impact = impact_text.replace('#gender_third', 'this person'),
        VOTE_STATUS = `${cosponsor_post_text}`;

    if (DID_SEARCH && NUMBER_REPRESENTATIVES === 1 && chamber === 'house') {
      const MEMBER_THIRD = REPRESENTATIVES[0].gender_full === 'man' ? 'He' : 'She',
            represent_text = represent.replace('#gender', MEMBER_THIRD);

      impact = impact_text.replace('#gender_third', `this ${REPRESENTATIVES[0].gender_full}`);
      VOTE_STATUS = REPRESENTATIVES[0].voted === 'Yea' ? ` ${single_voted_for} ${represent_text}` : ` ${single_voted_against} ${represent_text}`;

    } else if (DID_SEARCH && NUMBER_REPRESENTATIVES === 1 && chamber === 'senate'){
      VOTE_STATUS = REPRESENTATIVES[0].voted === 'Yea' ? ` ${cosponsor_post_text}` : '';
    }

    if (DID_SEARCH && NUMBER_REPRESENTATIVES === 1 && chamber === 'house' || DID_SEARCH && NUMBER_REPRESENTATIVES > 0 && chamber === 'senate') {
      this._initializeFullpage();
    } else {
      VOTE_STATUS = 'You have not yet searched for a member';
      this._destroyFullpage();
    }

    const blockClasses = cx(
      ['block', 'one'],
      {'hide': DID_SEARCH && NUMBER_REPRESENTATIVES === 1 && chamber === 'house' || DID_SEARCH && NUMBER_REPRESENTATIVES > 0 && chamber === 'senate'},
    );

    const backgroundClasses = cx(
      ['second-wrapper'],
      {'move-up': DID_SEARCH},
      {'static': !DID_SEARCH}
    );

    const sectionClasses = cx(
      ['section-block'],
      {'hide': NUMBER_REPRESENTATIVES === 1 || chamber === 'senate'}
    );

    const containerClasses = cx(
      ['container'],
      {'reveal': DID_SEARCH},
      {'green': !DID_SEARCH || NUMBER_REPRESENTATIVES > 1 || NUMBER_REPRESENTATIVES === undefined},
      {'orange': DID_SEARCH && NUMBER_REPRESENTATIVES !== 0 && NUMBER_REPRESENTATIVES !== undefined},
      {'red': DID_SEARCH && NUMBER_REPRESENTATIVES === 0 && chamber === 'senate'},
      {'visible': DID_SEARCH && NUMBER_REPRESENTATIVES === 0 && chamber === 'senate'},
      {'purple': this.state.current_screen === 2},
      {'full': DID_SEARCH}
    );

    return <div className={containerClasses}>
      <WhiteBorder />
      <div className="overlay">
        This site is only supported in portrait mode. Please turn your phone.
      </div>
      <div className={blockClasses} onScroll={this._handleScroll}>
      	<div className="section-block">
        	<Circle
          	style="one"
          	hide={true}
          	did_search={DID_SEARCH}
          	desc={bill_desc.replace('#member', MEMBER)}
        	/>
        	<SearchGroup
            repNum={NUMBER_REPRESENTATIVES}
          	error={ERROR}
            did_search={DID_SEARCH}
            zip_code={ZIP_CODE}
            state_full={STATE_FULL}
        	/>
        </div>
        <Results
          representatives={REPRESENTATIVES}
          numRep={NUMBER_REPRESENTATIVES}
          backgroundClasses={backgroundClasses}
          vote_status={VOTE_STATUS}
          impact={impact}
          current_member={CURRENT_MEMBER}
          zip_code={ZIP_CODE}
        />
      </div>
    </div>;
  }
};

export default Home;
