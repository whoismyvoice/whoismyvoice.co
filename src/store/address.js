import {
  RECEIVE_ADDRESS,
  RECEIVE_GPS,
  RECEIVE_ZIP_CODE,
  RESET_CURRENT,
} from '../actions/types';

const initialState = {
  value: undefined,
  street: undefined,
  zipCode: undefined,
  latitude: undefined,
  longitude: undefined,
};

function handle(state = initialState, action) {
  const { type } = action;
  switch (type) {
    case RECEIVE_ADDRESS:
      return {
        ...state,
        street: action.address.replace(new RegExp(`(, )?${state.zipCode}`), ''),
        value: action.address,
      };
    case RECEIVE_GPS:
      return {
        ...state,
        latitude: action.latitude,
        longitude: action.longitude,
      };
    case RECEIVE_ZIP_CODE:
      return {
        ...state,
        zipCode: action.zipCode,
      };
    case RESET_CURRENT:
      return initialState;
    default:
      return state;
  }
}

export default handle;
