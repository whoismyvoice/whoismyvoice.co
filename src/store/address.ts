import { Action, ActionType } from '../actions/types';

export type AddressState = {
  value?: string;
  street?: string;
  zipCode?: string;
  latitude?: string;
  longitude?: string;
};

const initialState = {
  value: undefined,
  street: undefined,
  zipCode: undefined,
  latitude: undefined,
  longitude: undefined,
};

function handle(state: AddressState = initialState, action: Action) {
  switch (action.type) {
    case ActionType.RECEIVE_ADDRESS:
      if (state.zipCode === undefined) {
        throw new Error('Can not receive street address when zipCode not set.');
      }
      return {
        ...state,
        street: action.address.replace(new RegExp(`(, )?${state.zipCode}`), ''),
        value: action.address,
      };
    case ActionType.RECEIVE_GPS:
      return {
        ...state,
        latitude: action.latitude,
        longitude: action.longitude,
      };
    case ActionType.RECEIVE_ZIP_CODE:
      return {
        ...state,
        zipCode: action.zipCode,
      };
    case ActionType.RESET_CURRENT:
      return initialState;
    default:
      return state;
  }
}

export default handle;
