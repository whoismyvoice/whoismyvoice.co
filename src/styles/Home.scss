@import 'variables';
@import 'transitions';

h1 {
  margin-top: 0;
  margin-bottom: 0;
  font-size: 60px;
  font-weight: bold;
}

.container {
  @include transition(1s);
  z-index: 1;
	width: 100vw;
	height: 100vh;
	overflow: hidden;
	-webkit-overflow-scrolling: touch;

  &.visible {
    overflow: visible;
  }

	@media screen and (max-width: $iphone6) {
		height: 100vh;
		min-height: 100%!important;
	}

	@media screen and (max-width: $iphone6)
	and (-webkit-min-device-pixel-ratio: 2)
	and (device-aspect-ratio: 375/667) {
		height: 89.5vh;
	}

	@media screen and (max-width: $iphone5-min) {
		height: 99vh;
		min-height: 100%!important;
	}

	@media screen and (max-width: $iphone5-min)
	and (-webkit-min-device-pixel-ratio: 2)
	and (device-aspect-ratio: 40/71) {
		height: 87vh;
	}

	@media only screen
  and (min-device-width : 414px)
  and (max-device-width : 736px)
  and (orientation : portrait)
  and (-webkit-min-device-pixel-ratio : 3) {
    height: 89.5vh;
  }
}

.full {
  height: auto;
  min-height: 99vh;
  width: 100vw;
}

.reveal {
  @include transition(1s);
  z-index: 1;
  overflow: hidden;
  width: 100vw;
}
