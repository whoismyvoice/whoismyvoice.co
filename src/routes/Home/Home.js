import React, { Component, } from 'react';
import PropTypes from 'prop-types';
import { connect, } from 'react-redux';
import cx from 'classnames';

// Components
import CornerRibbon from '../../components/CornerRibbon';
import Results from '../../components/Results';
import SearchGroup from '../../components/Search/SearchGroup';
import TitleComponent from '../../components/TitleComponent';

// Assets
import '../../styles/Home.css';

export class Home extends Component {
  static propTypes = {
    didSearch: PropTypes.bool,
    numberHouse: PropTypes.number,
    numberRepresentatives: PropTypes.number,
  }

  render() {
    const {
      didSearch,
      numberHouse,
      numberRepresentatives,
    } = this.props;

    const blockClasses = cx(
      ['block', 'block--margin'],
      {'disappear': didSearch && numberHouse === 1 && numberRepresentatives > 2},
    );

    const fadingClasses = cx(
      ['fading-circle'],
      {'orange-bg': didSearch && numberHouse === 1 && numberRepresentatives > 2}
    );

    const containerClasses = cx(
      ['container'],
      {'reveal': didSearch},
      {'full': didSearch && numberHouse === 1 && numberRepresentatives > 2}
    );

    return (
      <div className={containerClasses}>
        <CornerRibbon
          didSearch={didSearch}
         />

        <div className={fadingClasses}></div>
        <div className="overlay">
          This site is only supported in portrait mode. Please turn your phone.
        </div>
        <div className={blockClasses}>
        	<div className="section-block">
            <TitleComponent
              front={true}
              classes="title-component--padding"
              desc={true}
            />
          	<SearchGroup />
          </div>
          <Results destroy={this._destroyFullpage} />
        </div>
      </div>
    );
  }
}

function mapStateToProps(state) {
  const { address, officials, } = state;
  return {
    didSearch: address.value !== undefined,
    numberRepresentatives: officials.ids.length,
    numberHouse: officials.ids.length - 2,
    representatives: officials.ids.map(id => officials.byId[id]),
  };
}

export default connect(mapStateToProps)(Home);