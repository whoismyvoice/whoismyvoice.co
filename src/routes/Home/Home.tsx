import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import cx from 'classnames';
import * as mixpanel from 'mixpanel-browser';

// Components
import Results from '../../components/Results';
import SearchGroup from '../../components/SearchGroup';
import { StarTitle } from '../../components/Title/StarTitle';
import { Contribution } from '../../models/Contribution';
import { Record as LegislatorType } from '../../models/Legislator';
import { State } from '../../store';
import { Dispatch } from '../../actions/types';
// Constants
import { ORGANIZATION, ORGANIZATION_DISPLAY } from '../../constants';
import { HOME_ROUTE } from '../../constants/mixpanel-events';
// Assets
import '../../styles/Home.scss';

interface Props {
  currentPage?: 1 | 2;
  didSearch: boolean;
  numberHouse: number;
  numberRepresentatives: number;
  contributions: Array<Contribution>;
  representatives: Array<LegislatorType>;
}

export function Home(props: Props): JSX.Element {
  useEffect(() => {
    mixpanel.track(...HOME_ROUTE);
  }, []);
  const {
    currentPage,
    didSearch,
    numberHouse,
    numberRepresentatives,
    contributions,
    representatives,
  } = props;
  useEffect(() => {
    document.body.classList[didSearch ? 'add' : 'remove']('orange-bg');
  }, [didSearch]);

  const blockClasses = cx(['block'], {
    disappear: didSearch && numberHouse === 1 && numberRepresentatives > 2,
    'page-one': currentPage === undefined || currentPage === 1,
    'page-two': currentPage === 2,
  });

  const containerClasses = cx(
    ['container'],
    { reveal: didSearch },
    { full: didSearch && numberHouse === 1 && numberRepresentatives > 2 }
  );

  // tslint:disable
  const templateString = `Did my representatives accept campaign contributions <span class="strike-out">from <%= organizationName %>?</span>`;
  // tslint:enable
  const templateData = { organizationName: ORGANIZATION_DISPLAY };
  return (
    <div className={containerClasses}>
      <div className="overlay">
        This site is only supported in portrait mode. Please turn your phone.
      </div>
      <div className={blockClasses}>
        <div className="section-block">
          <StarTitle
            templateData={templateData}
            templateString={templateString}
          />
          <SearchGroup />
        </div>
        <Results
          contributions={contributions}
          representatives={representatives}
        />
      </div>
    </div>
  );
}

function mapStateToProps(state: State): Props {
  const { address, contributions, officials, view } = state;
  const organizationContributions =
    contributions.byOrganization[ORGANIZATION] || [];
  return {
    currentPage: view.currentPage,
    didSearch: address.value !== undefined,
    numberRepresentatives: officials.legislators.length,
    numberHouse: officials.legislators.length - 2,
    contributions: organizationContributions,
    representatives: officials.legislators,
  };
}

function mapDispatchToProps(dispatch: Dispatch): {} {
  return {};
}

export default connect(mapStateToProps, mapDispatchToProps)(Home);
