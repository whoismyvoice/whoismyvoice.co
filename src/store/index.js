import thunkMiddleware from 'redux-thunk';
import { createLogger, } from 'redux-logger';
import { applyMiddleware, combineReducers, createStore, } from 'redux';

import address from './address';
import contributions from './contributions';
import officials from './officials';
import view from './view';

const loggerMiddleware = createLogger();

export default createStore(
  combineReducers({
    address,
    contributions,
    officials,
    view,
  }),
  {},
  applyMiddleware(
    thunkMiddleware,
    loggerMiddleware,
  )
);
