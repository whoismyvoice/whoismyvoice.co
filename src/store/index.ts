import thunkMiddleware from 'redux-thunk';
import { createLogger } from 'redux-logger';
import * as promiseMiddleware from 'redux-promise';
import { applyMiddleware, combineReducers, createStore } from 'redux';

import address, { AddressState } from './address';
import contributions, { ContributionsState } from './contributions';
import officials, { OfficialsState } from './officials';
import view, { ViewState } from './view';

export type State = {
  address: AddressState;
  contributions: ContributionsState;
  officials: OfficialsState;
  view: ViewState;
};

const loggerMiddleware = createLogger();

export default createStore(
  combineReducers({
    address,
    contributions,
    officials,
    view,
  }),
  {},
  applyMiddleware(thunkMiddleware, promiseMiddleware, loggerMiddleware)
);
