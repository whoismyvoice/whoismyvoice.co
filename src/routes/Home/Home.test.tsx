import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { Home } from './Home';
import store from '../../store';

jest.mock('mixpanel-browser');

const props = {
  didSearch: false,
  numberHouse: 0,
  numberRepresentatives: 0,
  contributions: [],
  representatives: [],
};

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <Provider store={store}>
      <Home {...props} />
    </Provider>,
    div
  );
  ReactDOM.unmountComponentAtNode(div);
});

it('renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Home {...props} />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
