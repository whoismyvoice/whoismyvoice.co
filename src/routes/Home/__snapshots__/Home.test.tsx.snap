// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="container"
>
  <div
    className="fading-circle"
  />
  <div
    className="overlay"
  >
    This site is only supported in portrait mode. Please turn your phone.
  </div>
  <div
    className="block block--margin"
  >
    <div
      className="section-block"
    >
      <div
        className="title-component uppercase"
      >
        <div
          className="three-stars"
        >
          <span>
            ★
          </span>
          <span>
            ★
          </span>
          <span>
            ★
          </span>
        </div>
        <div
          className="title-component__description"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Did my representatives accept campaign contributions <span class=\\"strike-out\\">from the NRA?</span>",
            }
          }
        />
      </div>
      <div
        className="animated fadeIn"
      >
        <form
          method="GET"
          onSubmit={[Function]}
        >
          <input
            className=""
            name="zipCode"
            onChange={[Function]}
            pattern="[0-9]*"
            placeholder="Enter Your Zip Code"
            type="text"
          />
          <div
            className="line-seperator"
          />
          <button
            className="text-button"
            onClick={[Function]}
            type="submit"
          >
            Continue
            <span
              className="text-button-border"
            />
          </button>
        </form>
      </div>
    </div>
    <div
      className="second-wrapper"
      id="fullpage"
    />
  </div>
</div>
`;
